homeassistant/components/ws66i/config_flow.py

```diff
@@ -50,22 +51,29 @@ def _sources_from_config(data):
     }
 
 
-async def validate_input(hass: core.HomeAssistant, input_data):
-    """Validate the user input allows us to connect.
+async def validate_input(
+    hass: core.HomeAssistant, input_data: dict[str, Any]
+) -> dict[str, Any]:
+    """Validate the user input.
 
     Data has the keys from DATA_SCHEMA with values provided by the user.
     """
     ws66i: WS66i = get_ws66i(input_data[CONF_IP_ADDRESS])
-    await hass.async_add_executor_job(ws66i.open)
-    # No exception. run a simple test to make sure we opened correct port
-    # Test on FIRST_ZONE because this zone will always be valid
-    ret_val = await hass.async_add_executor_job(ws66i.zone_status, FIRST_ZONE)
-    if ret_val is None:
+
+    def _verify_connection(ws66i: WS66i) -> bool:
+        """Verify a connection can be made to the WS66i."""
+        ws66i.open()
+        # No exception. Verify correct port was opened
+        # Test on FIRST_ZONE because this zone will always be valid
+        ret_val = ws66i.zone_status(FIRST_ZONE)
+
         ws66i.close()
-        raise ConnectionError("Not a valid WS66i connection")
 
-    # Validation done. No issues. Close the connection
-    ws66i.close()
+        return bool(ret_val)
+
+    is_valid: bool = await hass.async_add_executor_job(_verify_connection, ws66i)
+    if not is_valid:
+        raise ConnectionError("Not a valid WS66i connection")
```

Comment:
We prefer to raise a custom exception that inherits `HomeAssistantError` instead of using a built in exception. I'd suggest to use the exception `CannotConnect` for this purpose. If the device can raise `ConnectionError` we can catch that here and raise `CannotConnect`.